theseus/theseus_layer.py

```diff
@@ -93,3 +118,154 @@ def device(self) -> torch.device:
+                    differentiable_tensors,
+                    allow_unused=True,
+                )
+            ctx.k = len(differentiable_keys)
+            ctx.save_for_backward(*input_vals, *grad_sol, *optim_tensors)
+        return (*optim_tensors, info)
+
+    @staticmethod
+    @once_differentiable
+    def backward(ctx, *grad_outputs):
+        saved_tensors = ctx.saved_tensors
+        input_vals = saved_tensors[: ctx.n]
+        grad_sol = saved_tensors[ctx.n : ctx.n + ctx.k]
+        optim_tensors = saved_tensors[ctx.n + ctx.k :]
+        grad_outputs = grad_outputs[:-1]
+
+        bwd_objective = ctx.bwd_objective
+        bwd_optimizer = ctx.bwd_optimizer
+        epsilon = ctx.epsilon
+        input_keys = ctx.input_keys
+        differentiable_keys = ctx.differentiable_keys
+
+        # Update the optim vars to their solutions.
+        bwd_data = dict(zip(input_keys, input_vals))
+        sol_values = dict(zip(bwd_objective.optim_vars.keys(), optim_tensors))
+        bwd_data.update(sol_values)
+
+        # Add in gradient values.
+        grad_data = {
+            TheseusLayerDLMForward._dlm_epsilon: torch.tensor(epsilon)
+            .to(grad_outputs[0])
+            .reshape(1, 1)
+        }
+        for i, name in enumerate(bwd_objective.optim_vars.keys()):
+            grad_data[name + TheseusLayerDLMForward._grad_suffix] = grad_outputs[i]
+        bwd_data.update(grad_data)
+
+        # Solve backward objective.
+        bwd_objective.update(bwd_data)
+        bwd_optimizer.optimize()
+
+        differentiable_tensors = [bwd_data[k] for k in differentiable_keys]
+
+        # Compute gradients.
+        with torch.enable_grad():
+            grad_perturbed = torch.autograd.grad(
+                bwd_objective.error_squared_norm().sum(),
+                differentiable_tensors,
+                allow_unused=True,
+            )
+
+        grads = []
+        i = 0
+        for k in input_keys:
+            if i < len(differentiable_keys) and k == differentiable_keys[i]:
+                if grad_sol[i] is not None:
+                    grads.append((grad_sol[i] - grad_perturbed[i]) / epsilon)
+                else:
+                    grads.append(None)
+                i += 1
+            else:
+                grads.append(None)
+        nones = [None] * len(grads)
+        return (None, None, None, None, None, None, *nones, *grads)
+
+
+def _dlm_perturbation(optim_vars, aux_vars):
+    v = optim_vars[0]
+    g = aux_vars[0]
+    epsilon = aux_vars[1]
+    return epsilon.data * v.data - 0.5 * g.data
+
+
+def _instantiate_dlm_bwd_objective(objective):
+    bwd_objective = objective.copy()
+    epsilon_var = Variable(torch.ones(1, 1), name=TheseusLayerDLMForward._dlm_epsilon)
+    for name, var in bwd_objective.optim_vars.items():
+        grad_var = Variable(
+            torch.zeros_like(var.data), name=name + TheseusLayerDLMForward._grad_suffix
+        )
+        bwd_objective.add(
```

Comment:
Also flagging that we can vectorize this or use Taosha's cost function grouping current in 'experimental'.